Automattic/jetpack
# Podcast Player: Introduce React for universal rendering

Experiments with making the player in React to be used in both the frontend and the editor.

Most of the code is reused from #15034, just structured in components and using React APIs.

Todo:
- [x] API endpoint for getting podcast episodes (to be used in the editor)
- [x] wrapper for the editor that loads content and uses the player component
- [x] work with attributes in components
- [x] keyboard access (fix spacebar)
- [x] evaluate dependency weight effect
- [x] much more!
- [x] refactor frontend block initialization. ~utilize `FrontendManagement`?~

#### Changes proposed in this Pull Request:
* Refactor the block to completely use React in both the editor and the frontend

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
paYJgx-sH-p2

#### Testing instructions:
- Rebuild extensions
- Insert a player block (or more) into a post and test in both editor and the frontend
- All operations of the player should work as you expect.
- Playing/pausing a track by:
  - Play/Pause button in the player
  - Clicking an episode link (however, clicking the same link with CMD modifier must do nothing and open the episode in a new tab)
  - Navigating the focus to an episode link and
    - pressing <kbd>Enter</kbd>
    - pressing <kbd>Space</kbd>
- ~Icons for an active episode showing/hiding as expected~ This will be a follow-up
- Errors shown when accessing non-playable episode (simulate by blocking requests/going offline)
  - Errors disappearing once another episode is selected
  - Link inside an error works
- Player instances are accessible in `window.jetpackPodcastPlayers.playerInstances` - you can control it externally by using its public methods `.play()`, `.pause()`, `.togglePlayPause()` and `.selectTrack( i )` where `i` is an index of episode in the list. Counted from zero as we developers love.
- Run `localStorage.debug = '*'` and reload the preview. You should see a message in the console about constructing a player instance.

#### Proposed changelog entry for your changes:
* none
